nexus/src/app/sagas/instance_migrate.rs

```diff
@@ -129,145 +255,160 @@ async fn sim_allocate_propolis_ip(
     let params = sagactx.saga_params::<Params>()?;
     let opctx = crate::context::op_context_for_saga_action(
         &sagactx,
         &params.serialized_authn,
     );
 
+    let db_instance = &params.instance;
     let migration_id = sagactx.lookup::<Uuid>("migrate_id")?;
-    let dst_sled_id = params.migrate_params.dst_sled_id;
     let dst_propolis_id = sagactx.lookup::<Uuid>("dst_propolis_id")?;
-    let (instance_id, old_runtime) =
-        sagactx.lookup::<(Uuid, InstanceRuntimeState)>("migrate_instance")?;
-
-    // Allocate an IP address the destination sled for the new Propolis server.
-    let propolis_addr = osagactx
-        .datastore()
-        .next_ipv6_address(&opctx, dst_sled_id)
+    let updated_record = osagactx
+        .nexus()
+        .instance_set_migration_ids(
+            &opctx,
+            db_instance.id(),
+            db_instance,
+            InstanceMigrationSourceParams { dst_propolis_id, migration_id },
+        )
         .await
         .map_err(ActionError::action_failed)?;
 
-    let runtime = InstanceRuntimeState {
-        sled_id: dst_sled_id,
-        propolis_id: dst_propolis_id,
-        propolis_addr: Some(std::net::SocketAddr::new(
-            propolis_addr.into(),
-            12400,
-        )),
-        ..old_runtime
-    };
+    Ok(updated_record)
+}
+
+async fn sim_clear_migration_ids(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let db_instance =
+        sagactx.lookup::<db::model::Instance>("set_migration_ids")?;
 
-    // Collect the external IPs for the instance.
-    //  https://github.com/oxidecomputer/omicron/issues/1467
-    // TODO-correctness: Handle Floating IPs, see
-    //  https://github.com/oxidecomputer/omicron/issues/1334
-    let (snat_ip, external_ips): (Vec<_>, Vec<_>) = osagactx
-        .datastore()
-        .instance_lookup_external_ips(&opctx, instance_id)
+    // If this call to clear migration IDs failed, one of the following things
+    // must have happened:
+    //
+    // 1. Sled agent's view of the instance changed in a way that invalidated
+    //    this request. Sled agent is expected to push its own state updates on
+    //    these transitions and properly manage migration IDs when they occur.
+    // 2. Nexus failed to contact sled agent entirely.
+    // 3. Nexus talked to sled agent but then failed to reach CRDB when trying
+    //    to write back the instance state with the cleared IDs. (Note that
+    //    when clearing migration IDs, failing to update CRDB because the
+    //    current generation number is too far advanced is not actually treated
+    //    as a failure.)
+    //
+    // In case 1, the instance should already be updated properly (or will be
+    // updated properly soon), and in cases 2 and 3 there's nothing that can
+    // reliably be done (the error may not be transient), so just swallow all
+    // errors here (but warn that they occurred).
```

Comment:
What is the outcome here if we hit errors here that are not-transient? Is the migration stuck?

Reply:
I corrected the comment in 1af08673 and 5e45c221. 

In general, the error handling story in the instance update paths seems very fragile to me. ISTM to work like this:

- When Nexus wants to change an instance's state, it issues a call to sled agent that returns a new instance runtime state on success or a sled agent error on failure. The returned result type is a Progenitor client result that wraps either the updated runtime state or a sled agent-produced HTTP error code.
- Instance state change results are handled in `Nexus::handle_instance_put_result`. If a state change failed, this converts the Progenitor client result into an Omicron error type. That conversion preserves 503 Unavailable errors, turns any 400-level client error into 400 Bad Request, and turns everything else into 500 Internal Server Error. Any further information from the inner error type is lost.
- Sled agent has a wide array of descriptive internal error types. However, its Dropshot error handlers need to return an HTTP error, and sled agent's error conversion routine turns almost everything into a 500 error, except for Propolis errors that bear their own statuses, which statuses get preserved.

This makes it really hard to get sled agent to express a specific instance state change error to Nexus: sled agent turns everything into a 500 regardless of the semantics of the internal error it encountered, but even if it didn't, `handle_instance_put_result` will just push everything through the `From<progenitor_client::Error>` impl, which throws away most individual error codes.

There's at least one bug in the new migration code that stems from all of this: setting or clearing migration IDs is supposed to return a 400-level error on a generation number conflict, but sled agent ignores the semantics of the `instance::Error::Transition` return type and just turns it into a 500 instead. This can cause instances to get marked as failed when they shouldn't be. I've tried to fix this in 28d526dd.